lib/fss.ex

```diff
@@ -13,19 +13,95 @@ defmodule FSS do
 
   defmodule S3 do
     defmodule Config do
-      defstruct [:region, :access_key_id, :secret_access_key]
+      defstruct [
+        :access_key_id,
+        :region,
+        :secret_access_key,
+        :endpoint,
+        :token
+      ]
 
       @type t :: %__MODULE__{
-              region: String.t(),
               access_key_id: String.t(),
-              secret_access_key: String.t()
+              region: String.t(),
+              secret_access_key: String.t(),
+              endpoint: String.t() | nil,
+              token: String.t() | nil
             }
     end
 
     defmodule Entry do
-      defstruct [:url, :config]
+      defstruct [:bucket, :key, :port, :config]
+
+      @type t :: %__MODULE__{
+              bucket: String.t(),
+              key: String.t(),
+              port: pos_integer(),
+              config: Config.t()
+            }
+
+      def config_from_system_env() do
+        %FSS.S3.Config{
+          access_key_id: System.get_env("AWS_ACCESS_KEY_ID"),
+          secret_access_key: System.get_env("AWS_SECRET_ACCESS_KEY"),
+          region: System.get_env("AWS_REGION", System.get_env("AWS_DEFAULT_REGION")),
+          token: System.get_env("AWS_SESSION_TOKEN")
+        }
+      end
+
+      def parse(url, opts \\ []) do
+        opts =
+          Keyword.validate!(opts, config: nil)
+
+        uri = URI.parse(url)
+
+        if uri.scheme == "s3" do
+          case uri do
+            %{host: host, path: "/" <> key} when is_binary(host) ->
+              config =
+                opts
+                |> Keyword.fetch!(:config)
+                |> case do
+                  nil ->
+                    config_from_system_env()
+
+                  %Config{} = config ->
+                    config
+
+                  config when is_list(config) or is_map(config) ->
+                    struct!(config_from_system_env(), config)
+
+                  other ->
+                    raise ArgumentError,
+                          "expect configuration to be a %FSS.S3.Config{} struct, a keyword list or a map. Instead got #{inspect(other)}"
+                end
+                |> validate_config!()
+
+              {:ok, %__MODULE__{bucket: host, key: key, port: uri.port, config: config}}
+
+            %{host: nil} ->
+              {:error, ArgumentError.exception("host is required")}
+
+            %{path: nil} ->
+              {:error, ArgumentError.exception("path to the resource is required")}
+          end
+        else
+          {:error, ArgumentError.exception("only s3:// URIs are supported for now")}
+        end
```

Comment:
I think you can remove some nesting if you push this scenario up,

```elixir
case uri do
  %{scheme: "s3", host: host, ...} ->
    ...

  %{scheme: scheme} when scheme != "s3" ->
    ...

  %{host: nil} ->
    ...

  ...
end
```

it's a bit redundant, but I think less nesting would be a win.

Also, nitpick:

> ```elixir
> %{host: nil} ->
>   {:error, ArgumentError.exception("host is required")}
> ```

while technically it is _host_, conceptually it is _bucket_. Maybe worth changing?

This gave me an idea, I think it could simply be:

```elixir
case URI.parse(url) do
  %{scheme: "s3", host: host, path: "/" <> key} when is_binary(host) ->
    ...

  url ->
    {:error,
     ArgumentError.exception("expected s3://<bucket>/<key> URL, got: " <> URI.to_string(url))}
end
```

but I might be trying too hard, I think it's fine either way. :)

Reply:
> while technically it is host, conceptually it is bucket. Maybe worth changing?

Maybe it makes sense to follow the field naming which Polars itself is using for this?
https://github.com/pola-rs/polars/blob/main/polars/polars-io/src/cloud/glob.rs#L76-L85

Reply:
@wojtekmach @Qqwy thanks for the suggestions! I applied most of them :)

I think I'm following more or less the naming which Polars is using, but I think they have more concepts that don't apply here.